feat(ktable): toggle column visibility [khcp-11162]

kaiarrowood opened the pull request:
# Summary

<!-- Insert a description of the changes in the PR, along with a JIRA ticket reference, if applicable. -->
Update `KTable` to add ability to toggle the visibility of columns.
For [KHCP-11162](https://konghq.atlassian.net/browse/KHCP-11162).

![image](https://github.com/Kong/kongponents/assets/67973710/2639bcc4-dcaa-4a7b-add6-08ec65513cef)

![image](https://github.com/Kong/kongponents/assets/67973710/47e36e10-990a-459e-b5e7-013d8f782aa9)


## PR Checklist

* [x] **Conventional Commits** all commits follow the conventional commit standards [outlined in the main README](https://github.com/Kong/kongponents#committing-changes).
* [x] **Tests coverage:** test coverage was added for new features and bug fixes
* [x] **Docs:** includes a technically accurate README


[KHCP-11162]: https://konghq.atlassian.net/browse/KHCP-11162?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

adamdehaven commented:
suggestion: I think we need a header or something at the top of the popover so the user knows what they're doing with these checkboxes
![image](https://github.com/Kong/kongponents/assets/2229946/bdf9ed3e-0f67-4298-8fbb-97fd0642e6c7)

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+
+<style lang="scss" scoped>
+.k-table-column-visibility-menu {
+  margin-left: auto;
+
+  .apply-button {
+    margin-left: auto;
+    margin-right: auto;
```
```suggestion
  margin-left: var(--kui-space-auto, $kui-space-auto);

  .apply-button {
    margin-left: var(--kui-space-auto, $kui-space-auto);
    margin-right: var(--kui-space-auto, $kui-space-auto);
```

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+
+const handleApply = () => {
+  // pass by ref problems
+  emit('update:visibility', JSON.parse(JSON.stringify(visibilityMap.value)))
```
question: why do we need `JSON.parse(JSON.stringify(` here? Can't we just emmit the original object, or does that produce mutations?

portikM commented on docs/components/table.md:
````diff
@@ -474,6 +475,30 @@ Example headers array:
 </script>
 ```
 
+#### `allowHide`
````
```suggestion
#### allowHide
```

nit: I'm trying to bring all headers to one standard which is not to apply any styling to header text of any level

adamdehaven replied:
See my comment on the prop naming 😄

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+            :label="col.label"
+          />
+        </KDropdownItem>
+        <KDropdownItem>
+          <KButton
+            appearance="tertiary"
+            class="apply-button"
+            @click="handleApply"
+          >
+            Apply
+          </KButton>
+        </KDropdownItem>
```
Apply button doesn't need to be wrapped by KDropdownItem (makes it look weird)

![Screenshot 2024-04-05 at 10 21 29 AM](https://github.com/Kong/kongponents/assets/36751160/554dd65c-c83f-464e-bb76-dd552c336c98)

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+<template>
+  <div class="k-table-column-visibility-menu">
+    <KDropdown>
+      <KTooltip text="Show/Hide Columns">
```
suggestion: we probably want to make tooltip text a prop (possible localization considerations?). Also I like sentence case better (_Show/hide columns_) but that might be just a personal preference.

kaiarrowood replied:
This component won't be exposed. We would end up needing the prop added to `KTable` which I don't think we want. Thoughts @adamdehaven ?

adamdehaven replied:
I believe we have other hard-coded text already in Kongponents? If yes I'd say just leave it hard-coded so we don't have prop-drilling 2-3 levels deep

portikM replied:
Yes we do. Okay, since this component won't be exported it's okay to leave it as is for now (or would be an overkill to add a prop just for this).

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+          class="menu-button"
+        >
+          <template #icon>
+            <CogIcon />
```
question: did we decide it will the cog icon? I remember we were discussing using columns icon or something

adamdehaven replied:
Yes I think a different icon would be more ideal. Material Symbols has some column icons I believe we could add to `@kong/icons`

kaiarrowood replied:
Picking up the new icon can be addressed in a followup PR once we have the desired icon in `@kong/icon`

adamdehaven replied:
TBH I don't think this is ideal; if the table starts getting used before then, it's going to be confusing for Konnect users to see a random "settings" icon and then have it change a few days later. 

You should be able to get an icon pretty easily from Jason in `#ask-kong-design-system`

portikM replied:
^ Second that, let's use the proper icon from beginning - I don't think design will have strong opinions against the columns icon

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+    <KDropdown>
+      <KTooltip text="Show/Hide Columns">
+        <KButton
+          appearance="secondary"
```
suggestion: maybe we should use `tertiary` appearance here?

adamdehaven commented on docs/components/table.md:
```diff
@@ -438,6 +438,7 @@ Pass in an array of header objects for the table.
 | `key`              | string  | A unique key for the column                                                                                                                |
 | `label`            | string  | The label displayed on the table for the column                                                                                            |
 | `sortable`         | boolean | Enables or disables server-side sorting for the column (`false` by default)                                                                |
+| `allowHide`        | boolean | Enable show/hide for this column                                                                                                           |
```
nit: can we rename this prop to align with the other properties? 
```suggestion
| `hidable`        | boolean | Enable show/hide for this column                                                                                                           |
```

portikM replied:
I don't think "hidable" is a word. Besides, this is not a prop but a property within an object that a prop accepts

portikM replied:
Oh okay, let's make it `hidable` then!

adamdehaven commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+onBeforeMount(() => {
+  // initialize visibility state
+  visibilityMap.value = props.columns.reduce((acc, col: TableHeader) => {
+    acc[col.label] = props.visibilityPreferences[col.label] === undefined ? true : props.visibilityPreferences[col.label]
```
question: do we need to point to a clone of the object instead of the object itself?

kaiarrowood replied:
For which? Aren't props non-mutatable?

adamdehaven replied:
was thinking where it evals to, but may be fine

```ts
acc[col.label] = props.visibilityPreferences[col.label]
```

adamdehaven commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,91 @@
+</script>
+
+<style lang="scss" scoped>
+.k-table-column-visibility-menu {
```
@portikM should we be using `k-*` classes here since this component isn't externally exposed?

portikM replied:
I thought it will be externally exposed since we want to have control over where we put it.. am I wrong @kaiarrowood?

portikM replied:
In that case yes, don't use `k-*` prefixed classes in this component

adamdehaven commented on src/components/KTable/KTable.vue:
```diff
@@ -93,7 +99,7 @@
               <th
-                v-for="(column, index) in tableHeaders"
+                v-for="(column, index) in visibleHeaders"
                 :key="`k-table-${tableId}-headers-${index}`"
```
can we use the column instead of the index as the key?

kaiarrowood replied:
I will change this but it's old code. Realistically this could wait until the reskin to fix?

adamdehaven replied:
You can skip for now

adamdehaven commented on src/components/KTable/KTable.vue:
```diff
@@ -169,11 +175,11 @@
-                v-for="(value, index) in tableHeaders"
+                v-for="(value, index) in visibleHeaders"
                 v-bind="cellAttrs({ headerKey: value.key, row, rowIndex, colIndex: index })"
                 :key="`k-table-${tableId}-cell-${index}`"
```
we shouldn't be using index in the key

adamdehaven replied:
You can skip for now

adamdehaven commented on src/components/KTable/KTable.vue:
```diff
@@ -1040,6 +1066,18 @@ const getTestIdString = (message: string): string => {
+  }
+
+  visibleHeaders.value = newVisibleHeaders
+}, { deep: true, immediate: true })
```
question: why does this need to be `immediate`? I wouldn't think it could change on mount?

kaiarrowood replied:
Because `visibleHeaders` needs to be initialized on mount once `tableHeaders` value is set.

adamdehaven commented on src/components/KTable/KTable.vue:
```diff
@@ -1040,6 +1066,18 @@ const getTestIdString = (message: string): string => {
+  const newHeaders = newVals[1]
+  const newVisibleHeaders = newHeaders.filter((header: TableHeader) => newVisibility[header.label] !== false)
+
+  if (JSON.stringify(newVisibleHeaders) !== JSON.stringify(visibleHeaders.value)) {
+    emitTablePreferences()
+  }
+
+  visibleHeaders.value = newVisibleHeaders
```
question: should we set the value before emitting?

Also, if the objects are equal, do we still need to set the `visibleHeaders.value` again?

kaiarrowood commented:
> suggestion: I think we need a header or something at the top of the popover so the user knows what they're doing with these checkboxes. Also may make sense to have a Cancel button ![image](https://private-user-images.githubusercontent.com/2229946/319706829-bdf9ed3e-0f67-4298-8fbb-97fd0642e6c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMjU4OTAsIm5iZiI6MTcxMjMyNTU5MCwicGF0aCI6Ii8yMjI5OTQ2LzMxOTcwNjgyOS1iZGY5ZWQzZS0wZjY3LTQyOTgtOGZiYi05N2ZkMDY0MmU2YzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDVUMTM1OTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjA0MWY2OTIxNWI3MmNiNWRhOGJmZDMzMmRjOGYxMDY3NWNlYzJkYmM5YWFiOGU0MTJiMjkyNTgwZjcyM2ZiZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.zK8NzGuMZrYH4ebJgC9rIYcJDxn1E0byU8AO6UpYIws)

@adamdehaven I added a tooltip, I still think we should use a different icon like you suggested but that's easy enough to change whenever design picks one.

![image](https://github.com/Kong/kongponents/assets/67973710/b00f1c18-773b-483c-a55d-5f35324eb53b)

I've added support to "cancel" when the dropdown is closed without clicking `Apply`. It will reset at that point.

portikM commented on src/components/KTable/ColumnVisibilityMenu.vue:
```diff
@@ -0,0 +1,141 @@
+      <KTooltip text="Show/Hide Columns">
+        <KButton
+          appearance="secondary"
+          class="menu-button"
```
issue: since this button will essentially be rendered in KTable toolbar - where often times we will have KInput and/or KSelect, this button needs to have `size="large"` to render equal size next to those components. However it will be too big when it's the only thing in the toolbar which isn't ideal - so maybe this is something we can control with a prop? Something like `columnVisibilityButtonSize` - seems oddly specific and I know @adamdehaven will hate it (so do I) but I'm sure displaying medium size button next to KInputs will be ab issue down the line

adamdehaven replied:
If we have inputs in the toolbar regularly, then just make the button size large by default? It shouldn't "change" depending on the siblings; we want consistency

portikM replied:
^ Works for me

kaiarrowood replied:
![image](https://github.com/Kong/kongponents/assets/67973710/c4d68810-d67f-4193-b6c4-130399f5b6f1)

![image](https://github.com/Kong/kongponents/assets/67973710/7002ad2b-ac89-4bbf-ad3c-cbdc63a7e230)

adamdehaven replied:
(regardless of the button/icon size) IMO the button just kind of "floating" there by itself is a bit odd; e.g. there's nothing anchoring it to the table or showing that it's related

portikM reviewed (approved):
LGTM 🚀
